sequence_task_3/student/include/impl.h

```diff
@@ -8,45 +8,78 @@ class DynamicPlaylist
 {
 public:
     /** @todo Member traits */
+	using value_type = typename Container::value_type;
+	using reference = typename Container::reference;
+	using const_reference = typename Container::const_reference;
+	using iterator = typename Container::iterator;
+	using const_iterator = typename Container::const_iterator;
+	using difference_type = typename Container::difference_type;
+	using size_type = typename Container::size_type;
 
     /** @todo Iterators */
 
+	const_iterator begin() const {return m_tracklist.cbegin();}
+	const_iterator end() const {return m_tracklist.cend();}
+
     DynamicPlaylist() = default;
 
     /** @todo Constructor from two iterators*/
+	template <typename it>
+	DynamicPlaylist(it first, it last) {
+		m_tracklist.assign(first, last);
+	}
 
     /** @todo Set track as currently played from initializer */
     template<class... Args>
-    const Song_t& play(Args&&... songData);
+    const Song_t& play(Args&&... songData) {
+		return *m_tracklist.emplace(begin(), std::forward<Args>(songData)...);
+	}
 
     /** @todo Set track as currently played from lvalue */
-    const Song_t& play(const Song_t& song);
+    const Song_t& play(const Song_t& song) { return *m_tracklist.insert(begin(), song);}
 
     /** @todo Set track to be played next from initializer */
     template<class... Args>
-    void playNext(Args&&... songData);
+    void playNext(Args&&... songData) {
+		short pos = hasTracks() ? 1:0;
```

Comment:
that's simple, but hacky, I'd use some helper like "getNextTrack", which will return iterator
but don't operate with any magic constants, just increment the iterator

Reply:
also consider - if there is no tracks, fall back to play